kubectl/upload/main.go

```diff
@@ -0,0 +1,212 @@
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+type Config struct {
+	Verbose    bool
+	Path       string
+	Kubeconfig string
+	Resource
+}
+
+type Resource struct {
+	Kind       string
+	Name       string
+	Namespace  string
+	EncodedMd5 string
+	// TODO(bjb): clean this up when the CRD changes we should only pass around
+	// encodedMd5s and not hex checksums
+	Md5Checksum string
+}
+
+type KubernetesClient interface {
+	CreateResource(gvr schema.GroupVersionResource, namespace string, obj *unstructured.Unstructured, opts metav1.CreateOptions) (*unstructured.Unstructured, error)
+	WatchResource(gvr schema.GroupVersionResource, namespace string, opts metav1.ListOptions) (watch.Interface, error)
+}
+
+type realKubernetesClient struct {
+	dynamicClient dynamic.Interface
+}
+
+func (c *realKubernetesClient) CreateResource(gvr schema.GroupVersionResource, namespace string, obj *unstructured.Unstructured, opts metav1.CreateOptions) (*unstructured.Unstructured, error) {
+	return c.dynamicClient.Resource(gvr).Namespace(namespace).Create(context.Background(), obj, opts)
+}
+
+func (c *realKubernetesClient) WatchResource(gvr schema.GroupVersionResource, namespace string, opts metav1.ListOptions) (watch.Interface, error) {
+	return c.dynamicClient.Resource(gvr).Namespace(namespace).Watch(context.Background(), opts)
+}
+
+func main() {
+	var cfg Config
+
+	var cmd = &cobra.Command{
+		Use:   "upload [kind] [name]",
+		Short: "Upload a resource of a given Kind and Name to be built as a substratus container image",
+		Args:  cobra.ExactArgs(2),
+		Run: func(cmd *cobra.Command, args []string) {
+			cfg.Resource.Kind = args[0]
```

Comment:
`kubectl` convetions is to accept lowercase: `kubectl get pods` vs `kubectl get Pods`. As long as this works, I think we are good.

Reply:
Ahh yeah it felt a bit strange using `Notebook` on the command line. I'll fix that up 👍